modules/ipc/include/hephaestus/ipc/zenoh/action_server.h

```diff
@@ -0,0 +1,250 @@
+  if (is_running_) {
+    LOG(ERROR) << fmt::format("ActionServer (topic: {}): server is already serving one request.",
+                              topic_config_.name);
+    return { .status = ActionServerRequestStatus::REJECTED_ALREADY_RUNNING };
+  }
+
+  try {
+    const auto response = request_cb_(request);
+    if (response != ActionServerRequestStatus::ACCEPTED) {
+      return { .status = response };
+    }
+  } catch (const std::exception& ex) {
+    LOG(ERROR) << fmt::format("ActionServer (topic: {}): request callback failed with exception: {}.",
+                              topic_config_.name, ex.what());
+    return { .status = ActionServerRequestStatus::INVALID };
+  }
+
+  if (!request_consumer_.queue().tryPush(request)) {
+    // NOTE: this should never happen as the queue is unbound.
+    LOG(ERROR) << fmt::format("ActionServer (topic: {}): failed to push the job in the queue. THIS SHOULD "
+                              "NOT HAPPEN CHECK THE CODE.",
+                              topic_config_.name);
+    return { .status = ActionServerRequestStatus::REJECTED_ALREADY_RUNNING };
+  }
+
+  LOG(INFO) << fmt::format("ActionServer (topic: {}): request accepted.", topic_config_.name);
+  return { .status = ActionServerRequestStatus::ACCEPTED };
+}
+
+template <typename RequestT, typename StatusT, typename ReplyT>
+void ActionServer<RequestT, StatusT, ReplyT>::execute(const RequestT& request) {
+  is_running_ = true;
+  // NOTE: we create the publisher and the subscriber only if the request is successful.
+  // This has the limit that that some status updates could be lost if the pub/sub are still discovering each
+  // other, but has the great advantage that we do not risk receiving messages from other requests if our
+  // request is rejected. auto publisher = Publisher<StatusT>{ session_,
+  // internal::getStatusPublisherTopic(topic_config_) };
+  auto publisher = Publisher<StatusT>{ session_, internal::getStatusPublisherTopic(topic_config_) };
+
+  std::atomic_bool stop_requested{ false };
+  auto stop_service = Service<std::string, std::string>(
+      session_, internal::getStopServiceTopic(topic_config_), [&stop_requested](const std::string&) {
+        stop_requested = true;
+        return "";
+      });
+
+  const auto reply = [this, &request, &publisher, &stop_requested]() noexcept {
+    try {
+      return execute_cb_(request, publisher, stop_requested);
+    } catch (const std::exception& ex) {
+      LOG(ERROR) << fmt::format("ActionServer (topic: {}): execute callback failed with exception: {}.",
+                                topic_config_.name, ex.what());
+      return ReplyT{};
+    }
+  }();
+  const auto response_topic = internal::getResponseServiceTopic(topic_config_);
+  // TODO: if stop_requested == true, should we communicate somehow that we terminate cause stop was
+  // requested?
+  const auto client_response =
+      callService<ReplyT, ActionServerRequestResponse>(*session_, response_topic, reply);
```

Comment:
maybe with timeout?

Reply:
great point, I'll add it like for normal `callService`